File: dotnet/DesignPatterns/BehavioralDesignPatterns/ObserverDesignPattern/ObserverDesignPattern.Example2/StatisticsDisplay.cs
namespace ObserverDesignPattern.Example2;

public class StatisticsDisplay : IObserver
{
    private float maxTemp = float.MinValue;
    private float minTemp = float.MaxValue;
    private float tempSum = 0.0f;
    private int numReadings = 0;
    private ISubject _weatherStation;

    public StatisticsDisplay(ISubject weatherStation)
    {
        _weatherStation = weatherStation;
        _weatherStation.RegisterObserver(this);
    }

    public void Update(float temperature, float humidity, float pressure)
    {
        tempSum += temperature;
        numReadings++;

        if (temperature > maxTemp)
            maxTemp = temperature;

        if (temperature < minTemp)
            minTemp = temperature;

        Display();
    }

    public void Display()
    {
        float avgTemp = tempSum / numReadings;
        Console.WriteLine($"📊 Statistics - Avg: {avgTemp:F1}°C, " +
                         $"Max: {maxTemp:F1}°C, Min: {minTemp:F1}°C");
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ObserverDesignPattern/ObserverDesignPattern.Example2/src/StatisticsDisplay.ts
import { IObserver } from './IObserver';
import { ISubject } from './ISubject';

export class StatisticsDisplay implements IObserver {
  private _maxTemp: number = -Infinity;
  private _minTemp: number = Infinity;
  private _tempSum: number = 0;
  private _numReadings: number = 0;
  private readonly _weatherStation: ISubject;

  constructor(weatherStation: ISubject) {
    this._weatherStation = weatherStation;
    this._weatherStation.registerObserver(this);
  }

  update(temperature: number, _humidity: number, _pressure: number): void {
    this._tempSum += temperature;
    this._numReadings++;

    if (temperature > this._maxTemp) this._maxTemp = temperature;
    if (temperature < this._minTemp) this._minTemp = temperature;

    this.display();
  }

  display(): void {
    const avgTemp = this._tempSum / this._numReadings;
    console.log(
      `📊 Statistics - Avg: ${avgTemp.toFixed(1)}°C, Max: ${this._maxTemp.toFixed(1)}°C, Min: ${this._minTemp.toFixed(1)}°C`
    );
  }
}
